Custom Tools

shaunak99 opened the pull request:
Introduces a way to define and add custom tools to agent alongside MCPServers.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Introduced a unified tool management system, allowing discovery and execution of both MCP and internal tools via a new ToolManager.
  * Added support for internal tools, including a new "search_history" tool for searching conversation history.
  * CLI and API updated to display and handle both MCP and internal tools.
  * SaikiAgent now provides unified methods for tool execution and discovery from all sources.

* **Bug Fixes**
  * Improved error messages and documentation for tool execution and management.

* **Documentation**
  * Updated API docs to reflect unified tool management and new methods.

* **Tests**
  * Added comprehensive tests for ToolManager, internal tools configuration, and integration scenarios.

* **Chores**
  * Refactored codebase to use ToolManager instead of direct MCPManager references.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

shaunak99 commented on src/core/tools/types.ts:
```diff
@@ -0,0 +1,141 @@
+import { z } from 'zod';
```
The zod import is already correct - it's an external npm package, not a local file, so it doesn't need the .js extension. The coding guidelines about .js extensions only apply to local file imports, not npm packages.

shaunak99 commented on tools/math.ts:
```diff
@@ -0,0 +1,73 @@
+import { createTool } from '@truffle-ai/saiki';
+import { z } from 'zod';
```
Current import is intended for example purposes with the published package.

rahulkarajgikar reviewed (commented):
have reviewed about 60-70%

a lot of this current design in the core backend is good, and the idea of enforcing the tools in a specific location is good, avoiding relative paths (extremely painful)

but my hunch is that dynamically loading modules from tools/directory at runtime will be quite error prone and unreliable - i had similar problems while working on plugins PR, and the whole design ended up being super brittle and unreliable. some of the problems:
 - errors in code for custom tools won't be caught at compile time, because we never actually compile the code for the tool
 - incorrect tool names in YML also won't be caught early
 - this will lead to having to actually load up the agent each time, and testing out a specific prompt for the model that uses that tool, for something which could be caught at compile time. there are some ways to catch some of this earlier, but the primary issue here is that ts code should be getting compiled to js and validated as part of a step before the agent runs at all
 - dependence on current working directory to resolve tools is brittle and ideally shouldn't be there - it enforces users to run commands only from a specific location which can easily be missed by users (they walk up a directory, re-run CLI with updated filepaths, everything now breaks)
 
to solve these problems, have been working on this design which aims to solve a lot of these issues: 
https://www.notion.so/Saiki-revamped-CLI-design-23e63b89f2dd80d8b60eea69aa252ff5?pvs=12 - 

it introduces a fixed project structure + an intermediate `build` step which is responsible for handling custom code (like custom tools). this also gives us a clear way to think about the CLI (inside a saiki project v/s outside the saiki project)

we enforce users to get started with saiki create-app, enforce locations with CLI scaffold commands, and catch custom code issues at compile time with saiki build, kind of like next-js

instead of having cwd based file detection, we use project based file detection to understand if the cwd is inside a saiki project (check current path.ts code -> we recursively walk up from cwd until we hit a package.json that has a dependency on @ truffle-ai/saiki.)

this moves away from having a global CLI that can run anywhere, but gives us a much more robust useful CLI, solves the custom code problem, and gives users a well-defined project structure for them to experiment with agents

i am planning to move towards this design, lmk your thoughts.

we can likely re-use a lot of the code in this PR, but some of the modules will be completely different (mostly path resolution/tool building logic)

shaunak99 commented:
Appreciate the detailed feedback! Based on our discussion, I have removed the `customTools` system and added only the essential parts needed for internal-tools while preserving the `ToolManager` implementation. I have also tried to update the MCPManager and ToolManager to keep them as first-class services. Let me know if any changes need to be made here. 

Additionally, I have added an internal-tool using the `SearchService` to enable the agent to have the ability to search & reference past session interactions while chatting. Will be adding a similar service for scheduler.

rahulkarajgikar commented on src/core/tools/types.ts:
```diff
@@ -0,0 +1,57 @@
+/**
+ * Tool parameters interface for normalized tool definitions
+ */
+export interface ToolParameters {
+    type: 'object';
+    properties: Record<string, unknown>;
+    required?: string[];
+    [key: string]: unknown; // Allow additional JSON Schema properties
+}
+
+/**
+ * Raw tool definition interface (before normalization)
+ */
+export interface RawToolDefinition {
+    description?: string;
+    parameters?: {
+        type?: string;
+        properties?: Record<string, unknown>;
+        required?: string[];
+    };
+}
+
+/**
+ * Tool manager's unified tool set (for compatibility with MCP tools and ai/types.ts)
+ */
+export interface ToolManagerToolSet {
+    [toolName: string]: {
+        name?: string;
+        description: string;
+        parameters?: ToolParameters;
+    };
+}
```
we already have most of these types in this file in core/ai/types.ts

we can move those types into here and add the new ones, these are mostly duplicate

there are also some small subtle changes in these duplicate types that are leading to a whole lot of unnecessary code - `ToolManagerToolSet` is the same as `ToolSet` except that description is no longer optional.

this is resulting in `ToolManager.getMcpTools()`, which should be just one line calling `McpManager.getAllTools()` needing unnecessary type conversion between the two ToolSet types. just using ToolSet everywhere would have avoided this

there are some other problems, can we make a gh issue for this, i will fix in follow up

rahulkarajgikar commented on src/core/tools/types.ts:
```diff
@@ -0,0 +1,57 @@
+    description: string;
+
+    /** Zod schema defining the input parameters */
+    inputSchema: any; // Using any for flexibility since we don't import zod here
```
import zod here and don't use any

rahulkarajgikar commented on src/core/utils/service-initializer.ts:
```diff
@@ -94,30 +99,53 @@ export async function createAgentServices(
+    // 5. Initialize search service
+    const searchService = new SearchService(storage.database);
+
+    // 6. Initialize internal tools provider with services and confirmation support
+    const internalToolsProvider = new InternalToolsProvider(
+        {
+            searchService,
+            // Future services can be added here as needed
+        },
+        confirmationProvider,
+        config.internalTools
+    );
```
can be internal dependency inside ToolManager

rahulkarajgikar reviewed (approved):
tested it out, works well!

there are some duplicate types and unnecessary code in the tools module which can be simplified, i'll merge this and fix it in a follow up
